sdkproject/Assets/Mapbox/Examples/Scripts/QuadTreeCameraMovement.cs

```diff
@@ -0,0 +1,172 @@
+﻿namespace Mapbox.Examples
+{
+	using Mapbox.Unity.Map;
+	using Mapbox.Unity.Utilities;
+	using Mapbox.Utils;
+	using UnityEngine;
+	using System;
+
+	public class QuadTreeCameraMovement : MonoBehaviour
+	{
+		[SerializeField]
+		[Range(1, 20)]
+		public float _panSpeed = 1.0f;
+
+
+		[SerializeField]
+		public Camera _referenceCamera;
+
+		[SerializeField]
+		QuadTreeTileProvider _quadTreeTileProvider;
+
+		[SerializeField]
+		AbstractMap _dynamicZoomMap;
+
+		private Vector3 _origin;
+		private Vector3 _mousePosition;
+		private Vector3 _mousePositionPrevious;
+		private bool _shouldDrag;
+
+		void Start()
+		{
+			if (null == _referenceCamera)
+			{
+				_referenceCamera = GetComponent<Camera>();
+				if (null == _referenceCamera) { Debug.LogErrorFormat("{0}: reference camera not set", this.GetType().Name); }
+			}
+		}
+
+
+		private void LateUpdate()
+		{
+			if (null == _dynamicZoomMap) { return; }
+
+			if (Input.touchSupported)
+			{
+				HandleTouch();
+			}
+			else
+			{
+				HandleMouseAndKeyBoard();
+			}
+		}
+
+		void HandleMouseAndKeyBoard()
+		{
+			// zoom
+			float scrollDelta = 0.0f;
+			scrollDelta = Input.GetAxis("Mouse ScrollWheel");
+			ZoomMapUsingTouchOrMouse(scrollDelta);
+
+			//pan keyboard
+			float xMove = Input.GetAxis("Horizontal");
+			float zMove = Input.GetAxis("Vertical");
+
+			PanMapUsingKeyBoard(xMove, zMove);
+
+
+			//pan mouse
+			PanMapUsingTouchOrMouse();
+		}
+
+		void HandleTouch()
+		{
+			float zoomFactor = 0.0f;
+			//pinch to zoom. 
+			switch (Input.touchCount)
+			{
+				case 1:
+					{
+						PanMapUsingTouchOrMouse();
+					}
+					break;
+				case 2:
+					{
+						// Store both touches.
+						Touch touchZero = Input.GetTouch(0);
+						Touch touchOne = Input.GetTouch(1);
+
+						// Find the position in the previous frame of each touch.
+						Vector2 touchZeroPrevPos = touchZero.position - touchZero.deltaPosition;
+						Vector2 touchOnePrevPos = touchOne.position - touchOne.deltaPosition;
+
+						// Find the magnitude of the vector (the distance) between the touches in each frame.
+						float prevTouchDeltaMag = (touchZeroPrevPos - touchOnePrevPos).magnitude;
+						float touchDeltaMag = (touchZero.position - touchOne.position).magnitude;
+
+						// Find the difference in the distances between each frame.
+						zoomFactor = 0.01f * (touchDeltaMag - prevTouchDeltaMag);
+					}
+					ZoomMapUsingTouchOrMouse(zoomFactor);
+					break;
+				default:
+					break;
+			}
+		}
+
+		void ZoomMapUsingTouchOrMouse(float zoomFactor)
+		{
+			_quadTreeTileProvider.UpdateMapProperties(_dynamicZoomMap.CenterLatitudeLongitude, Mathf.Max(0.0f, Mathf.Min(_dynamicZoomMap.Zoom + zoomFactor * 0.25f, 21.0f)));
```

Comment:
What are those constants?
`0.25` and `21`?

Reply:
0.25 is the factor by which zoom is incremented/ decremented on mouse scroll.
21 is the highest zoom level for the tileservice

Reply:
Should these be serialize fields?

Reply:
Zoom factor should be serializable, highest and lowest zoom levels should probably come from the service.